### ISSUE TITLE
Crash when running plan on VMs with multiple network interfaces with no gateway set

### ISSUE THREAD

**Issue by icksa** [terraform-providers/terraform-provider-vsphere#173]
Hello:

I am running into a terraform crash that I can't seem to pin down or work around. I am creating VMs in vsphere. The VMs have two network interfaces attached one of which has no gateway defined. Creating the VM initially is no problem, but running subsequent `terraform plan` commands are producing terraform crashes. The crash.log is here:

https://gist.github.com/icksa/a87c295aa976bcd805981e2be16c03b8#file-crash-log

There terraform file is here:

https://gist.github.com/icksa/a87c295aa976bcd805981e2be16c03b8#file-terraform-tf

The relevant section of the crash logs appears to be:
```
2017-09-21T09:28:16.960-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 2017/09/21 09:28:16 [DEBUG] networkInterface: map[string]interface {}{"key":4000, "ipv4_address":"10.200.96.51", "ipv4_prefix_length":20, "label":"Stratus6", "mac_address":"00:50:56:9a:12:f0"}
2017-09-21T09:28:16.960-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 2017/09/21 09:28:16 [DEBUG] networks: []map[string]interface {}{map[string]interface {}{"ipv4_address":"10.200.96.51", "ipv4_prefix_length":20, "label":"Stratus6", "mac_address":"00:50:56:9a:12:f0", "key":4000}, map[string]interface {}{"label":"Stratus3", "mac_address":"00:50:56:9a:a0:5e", "key":4001, "ipv4_address":"10.200.20.71", "ipv4_prefix_length":22}}
2017-09-21T09:28:16.960-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 2017/09/21 09:28:16 [DEBUG] ipv4_gateway of device id 5: 10.200.20.5
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: panic: runtime error: index out of range
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4:
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: goroutine 22 [running]:
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.resourceVSphereVirtualMachineRead(0x4233d7570, 0x1e8b7e0, 0x42337a8d0, 0x1, 0x29d3aa0)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:1180 +0x46a7
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Refresh(0x423467740, 0x4233ea370, 0x1e8b7e0, 0x42337a8d0, 0x423387508, 0x423469201, 0x80000000018)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:321 +0x199
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Refresh(0x423207e30, 0x4233ea320, 0x4233ea370, 0x8422fb4d90, 0x0, 0x18)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:284 +0x9a
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Refresh(0x423351a20, 0x42311cde0, 0x42311d050, 0x0, 0x0)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:510 +0x4e
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: reflect.Value.call(0x42334f200, 0x4233ec640, 0x13, 0x1e9e540, 0x4, 0x422ffcf20, 0x3, 0x3, 0x0, 0x0, ...)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/reflect/value.go:434 +0x906
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: reflect.Value.Call(0x42334f200, 0x4233ec640, 0x13, 0x4233e6f20, 0x3, 0x3, 0x423307020, 0x0, 0x0)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: net/rpc.(*service).call(0x4233aadc0, 0x4234d5400, 0x4233d4710, 0x4233dcc80, 0x423351f80, 0x1b635c0, 0x42311cde0, 0x16, 0x1b63600, 0x42311d050, ...)
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: created by net/rpc.(*Server).ServeCodec
2017-09-21T09:28:16.963-0500 [DEBUG] plugin.terraform-provider-vsphere_v0.3.0_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2017/09/21 09:28:16 [ERROR] root: eval: *terraform.EvalRefresh, err: vsphere_virtual_machine.atl-ch3-vm-mesosworker01: unexpected EOF
2017/09/21 09:28:16 [ERROR] root: eval: *terraform.EvalSequence, err: vsphere_virtual_machine.atl-ch3-vm-mesosworker01: unexpected EOF
2017/09/21 09:28:16 [TRACE] [walkRefresh] Exiting eval tree: vsphere_virtual_machine.atl-ch3-vm-mesosworker01
2017/09/21 09:28:16 [TRACE] dag/walk: upstream errored, not walking "provider.vsphere (close)"
2017-09-21T09:28:16.965-0500 [DEBUG] plugin: plugin process exited: path=/Users/milad/git/atlas-terraform/atlas-prod/ch3/vms/mesos-worker/.terraform/plugins/darwin_amd64/terraform-provider-vsphere_v0.3.0_x4
2017/09/21 09:28:16 [DEBUG] plugin: waiting for all plugin processes to complete...
2017-09-21T09:28:16.965-0500 [WARN ] plugin: error closing client during Kill: err="connection is shut down"
```

I am not sure exactly what is causing this but I would greatly appreciate any help. This crash is happening pretty regularly now...Please let me know what additional information to send.

Thanks

**Comment by icksa** [comment]
Is this the right forum for this question...should I post on the terraform project instead?

**Comment by vancluever** [comment]
Hey @icksa, sorry you are having trouble and the delay in replying!

This is indeed the correct place for this question - I am reviewing now and will probably have some questions in a bit. Cheers!

**Comment by vancluever** [comment]
Hey @icksa - got a successful reproduction over here - will probably get a fix in later today. Sorry you experienced this!

If you're curious - the root cause is due to the fact that there seems to be a large number of software interfaces on this system (versus virtual NICs set up in vSphere). The guest network discovery code (which will probably largely be going away during the pending rewrite of the resource) tries to discover what interface the gateway belongs to so that it can properly calculate diffs. It does this by taking the `device` index on the route setting, which would normally be a 1-to-1 mapping of the devices actually on this system. However, the provider only counts for state the NICs that actually have device equivalents in vSphere (in this case, there is a device ID that will correspond with a specific NIC in the actually VM configuration).

What's happening here is since VMware tools has discovered interfaces _before_ the ones actually configured in vSphere, which has thrown the interface index off, and as a result it's trying to save the gateway information in device index `5` when there are only two interfaces that have been accounted for on the system, causing the range error.

I was able to reproduce this with a simple TF config where I just added a dummy interface after the fact and set the gateway to route through that interface, so I should be able to fix it quick now that I have a stable repro.

Thanks and watch this space!

**Comment by vancluever** [comment]
Hey @icksa - if you have the ability to do so, you can try again with a [custom built provider binary](https://github.com/terraform-providers/terraform-provider-vsphere#building-the-provider) against the branch referenced in the above pull request?

This fixes my repro, but I want to make sure that fixes your crash.

Thanks!

**Comment by icksa** [comment]
Hi @vancluever: I'll give it a shot

**Comment by icksa** [comment]
@vancluever, sorry for the delay...I just tried it out and this works for me so please go ahead and merge. If you merge do you know approximately how long it will be before released? Thanks so much, this is tremendously helpful.

**Comment by vancluever** [comment]
@icksa - I'm hoping to get a release out later this week, mainly after merging this and getting a updated govmomi in to fix issues like #161 and #165.

Thanks and watch this space!
